### PR TITLE
feature:(azure_blob sink): Add support for Azure Blob Storage Endpoint Override

### PR DESCRIPTION
**Pull request by archoversight**
This allows setting the Azure Blob Storage Endpoint to a custom URL so that we may use the automated credentials within alternate clouds such as USGov and China.

This is required because the Azure SDK for Rust naively assumes that it is always running in the Public Azure, which uses `.core.windows.net` as the base URL suffix, which is not correct in USGovernment environments of Azure where that should be: `core.usgovcloudapi.net`.

I have opened an feature request with the Azure SDK for Rust: https://github.com/Azure/azure-sdk-for-rust/issues/1156 to automatically detect which cloud it is running in, but in the meantime this will provide an override that allows it to bet set manually when wanting to to use the machine identity to be able to connect to blob storage vs using the `connection_string` (which already had support for overriding the blob endpoint).

### CODE REVIEW

File: src/sinks/azure_common/config.rs
Comment (archoversight) [inline src/sinks/azure_common/config.rs:136]: I refactored this code to remove a need for a mutable builder, and to avoid creation of a CloudLocation and then throwing it away by overwriting it with `build.cloud_location`.

File: src/sinks/azure_common/config.rs
Comment (archoversight) [inline src/sinks/azure_common/config.rs:153]: Technically `storage_account_p` is not used when a `blob_endpoint` is, due to the way that the `CloudLocation` works, but I didn't want to add additional logic for managing a third case and duplicate the code necessary here for the storage_credentials setup.

I can if necessary though and make `blob_endpoint` mutually exclusive with `storage_account` and `connection_string`.

**Comment by archoversight** [comment]
I forgot to fix the tests. Will do.

**Comment by archoversight** [comment]
What formatter do I use to format the cue files automatically?

**Comment by archoversight** [comment]
```
target/debug/vector generate-schema > /tmp/vector-config-schema.json 2>/dev/null
scripts/generate-component-docs.rb /tmp/vector-config-schema.json \

scripts/generate-component-docs.rb:1494:in `<main>': undefined method `filter_map' for #<Hash:0x0000000100918128> (NoMethodError)
Did you mean?  filter
make: *** [generate-component-docs] Error 1
```

Trying to follow instructions here: https://github.com/vectordotdev/vector/actions/runs/3569094895/jobs/5998691094#step:16:28

File: src/sinks/azure_common/config.rs
Comment (archoversight) [inline src/sinks/azure_common/config.rs:138]: When a `blob_endpoint` is specified in the connection string, the `account_name` does not need to be specified in the connection string.

**Review by davidhuie-dd — CHANGES_REQUESTED** [review]
A few minor suggestions, but this looks good!

**Comment by davidhuie-dd** [comment]
@archoversight Another thought, is that y'all could use the this Vector fork while the upstream Azure change is made. When that's done, we could just upgrade the Azure dep.

**Comment by archoversight** [comment]
@davidhuie-dd unfortunately it is unlikely that this issue is going to get fixed upstream anytime soon. The Azure Rust SDK makes far too many assumptions about it being run inside of the commercial location and does not adequately support alternate locations when it comes to automatically adding the right postfix to the account name. Let me provide some additional background if I may:

Even if the Azure SDK would correctly support providing the cloud location, Vector would need to be updated to use the right enum depending on the cloud it is in, which is currently:

```
% az cloud list --output table
IsActive    Name               Profile
----------  -----------------  ---------
False       AzureCloud         latest
False       AzureChinaCloud    latest
True        AzureUSGovernment  latest
False       AzureGermanCloud   latest
```

At the moment the following are supported by Azure SDK for Rust:

- Public
- China
- Emulator

Or 

- Custom

See here: https://github.com/Azure/azure-sdk-for-rust/blob/3de510ce055be3cdc65b3b46783d4f6d97db4fe7/sdk/storage/src/cloud_location.rs#L9-L27

So either Vector gets in the business of writing code each time Azure adds a new region and switches the enum used for the CloudLocation or we end up just using the URL that the user passes `blob_endpoint` manually and we instantiate the `Custom` enum instead.

Unfortunately Azure does not provide a way (that I have found) for a system to self-identity using the `ManagedIdentityCredential`'s which of the above clouds it is in so that for example the script that starts Vector would be able to set the right Azure cloud programmatically.

The `blob_endpoint` can be set in the `connection_string` and in fact that is required when operating outside of the commercial Azure cloud (`AzureCloud`) because otherwise the access key/SAS that is provided in the connection string will not work as once again the Azure SDK for Rust will postfix the `account_name` with the wrong URL.

However we can't use the `connection_string` with the standard credential providers:

- EnvironmentCredential
- ManagedIdentityCredential
- AzureCliCredential

Because Vector assumes that the connection string contains all the credentials required and does not attempt the alternate credential providers.

We could refactor things in Vector so that we can use `connection_string` but not assume that it contains a access key/SAS and attempt the standard credential providers while still allowing a user to override the various Azure SDK parameters by passing them as semi-colon separated key value pairs. I'd be willing to do that work too.

I apologize, this may have been a long plea to "Please don't wait on the Azure SDK to fix this".

**Comment by archoversight** [comment]
@davidhuie-dd thank you so much for your review. I've added your suggestions into the code, and rebased on top of `master`.

File: src/sinks/azure_blob/config.rs
Comment (bruceg) [inline src/sinks/azure_blob/config.rs:62]: The common convention for a config parameter like this would be simply `endpoint`. I think the other `blob_` prefixed names are specific to the individual blob stored (ie filename modifiers).

**Review by bruceg — COMMENTED** [review]
The code looks good and I appreciate the refactoring. I have left a comment about the setting naming though.

File: src/sinks/azure_blob/config.rs
Comment (archoversight) [reply src/sinks/azure_blob/config.rs:62]: I'm okay with renaming it to `endpoint` but it feels somewhat weird because Azure has many endpoints. This one is specific for blob storage and matches the name you'd find in Azure documentation. It is also the name of the k/v that is passed in to the `connection_string` when using a SAS/access key.

https://learn.microsoft.com/en-us/connectors/azureblob/#get-the-azure-storage-blob-endpoint-for-a-given-storage-account

Comment (davidhuie-dd) [reply src/sinks/azure_blob/config.rs:62]: I'll defer to @bruceg here, but the context and namespace here is Azure blob storage, so it does feel redundant.

**Review by davidhuie-dd — COMMENTED** [review]
LGTM, but I'm waiting on @bruceg's comments before approving.

File: src/sinks/azure_blob/config.rs
Comment (bruceg) [reply src/sinks/azure_blob/config.rs:62]: That was my thinking as well. Any other Azure endpoints would be served by a different _type_ of sink, and we are already in the context of the "blob" service. In the same way, we have many different types of AWS and GCP sinks, but they all have just an optional `endpoint` option (albeit there with the complication of a `region` setting for AWS, but that's a separate matter) rather than a `s3_endpoint`, `gcs_endpoint`, etc.

Comment (archoversight) [reply src/sinks/azure_blob/config.rs:62]: Alright! I will make the change to name. Will push that either a little later today or tomorrow morning.

Thanks!

Comment (archoversight) [reply src/sinks/azure_blob/config.rs]: Things got away from me last week, but it should all be resolved now!

**Comment by archoversight** [comment]
@davidhuie-dd @bruceg I've updated everything with the latest comments + rebased against the latest changes on the main branch.

I am going to be on vacation until January, I would love to see this get merged if possible. If more changes are needed I won't be able to get to them. I did enable the feature that allows maintainers of vector to push to my branch on my fork, so if its is smaller minor nits please feel free make the change.

I appreciate all you do, and thanks for helping get this past the finish line!

File: src/sinks/azure_blob/config.rs
Comment (archoversight) [reply src/sinks/azure_blob/config.rs]: Updated.

**Comment by bruceg** [comment]
This will need to merge with `master` once #15797 lands, as the VRL tests are currently broken.

**Comment by archoversight** [comment]
Rebased on top of `master`.

**Comment by archoversight** [comment]
When I run this:

https://github.com/vectordotdev/vector/actions/runs/3840504621/jobs/6539649503#step:16:1

Locally I get 0 files changed... I am not sure how to please this check.

**Comment by bruceg** [comment]
I ran it myself and it regenerated the file, so I pushed it up here.

**Review by bruceg — APPROVED** [review]

**Comment by archoversight** [comment]
In an attempt to make sure this doesn't go stale, is there anything additional that I can do to help get this merged?

Thanks!

**Review by davidhuie-dd — APPROVED** [review]
thanks for this!

**Pull request merged**
